vllm/distributed/device_communicators/shm_broadcast.py

```diff
@@ -0,0 +1,119 @@
+import time
+import multiprocessing
+from multiprocessing import shared_memory, resource_tracker
+from multiprocessing import Array, Value
+import contextlib
+from contextlib import contextmanager
+from torch.distributed import ProcessGroup
+import torch
+import torch.distributed as dist
+import pickle
+
+
+class ShmRingBuffer:
+
+    def __init__(self, pg: ProcessGroup, max_chunk_bytes, max_chunks):
+        self.rank = dist.get_rank(pg)
+        self.world_size = dist.get_world_size(pg)
+        global_ranks = dist.get_process_group_ranks(pg)
+        self.is_writer = self.rank == 0
+        self.is_reader = not self.is_writer
+        self.current_idx = 0
+        self.max_chunk_bytes = max_chunk_bytes
+        self.max_chunks = max_chunks
+        total_bytes = (self.max_chunk_bytes +
+                       self.world_size) * self.max_chunks
+        self.buffer_offset = 0
+        self.metadata_offset = self.max_chunk_bytes * self.max_chunks
+
+        if self.is_writer:
+            self.shared_memory = shared_memory.SharedMemory(create=True,
+                                                            size=total_bytes)
+            # initialize the buffer to 0
+            for i in range(total_bytes):
+                self.shared_memory.buf[i] = 0
```

Comment:
why do we need to do this?

Reply:
this is to make sure the metadata section has correct initial state.

Reply:
can we just limit it to the metadata section? this can add overhead with large buffers

Reply:
fixed in [ba6839d](https://github.com/vllm-project/vllm/pull/5399/commits/ba6839d2aad668496402bc7ba4fd85cc51ff9c5a)